### PR TITLE
Add the pinned attribute to the alloca op

### PR DESCRIPTION
**Pull request by kiranchandramohan**
This patch has the following changes,
1) A new attribute `pinned` was added to the allocaOp. This attribute says that the alloca has already been hoisted to the right position and need/should not be moved again. These allocas are usually temporaries that were hoisted (https://github.com/flang-compiler/f18-llvm-project/blob/51261bd85f925327bccf1badf6befba1a02cd530/flang/lib/Optimizer/Builder/FIRBuilder.cpp#L204) or in OpenMP,  the privatised variables.
2) More variants of the builder functions and allocateLocal function were added to have functions to build/create allocaOp with the pinned attribute.
3) Several testcases were modified to add the pinned attribute. I think these are all temporaries that were created.

This change was originally requested in https://github.com/flang-compiler/f18-llvm-project/pull/928#discussion_r682915810. Pulling it out (with modifications) for ease of review, approval and merge.

### CODE REVIEW

File: flang/test/Lower/array-constructor-2.f90 [hunk 1]
```diff
@@ -104,8 +104,8 @@ subroutine test4(a, b, n1, m1)
   integer, external :: f1, f2, f3
 
   !  Dynamic array ctor with dynamic extent using implied do loops.
-  ! CHECK-DAG: fir.alloca index {uniq_name = ".buff.pos"}
-  ! CHECK-DAG: fir.alloca index {uniq_name = ".buff.size"}
+  ! CHECK-DAG: fir.alloca index {pinned, uniq_name = ".buff.pos"}
+  ! CHECK-DAG: fir.alloca index {pinned, uniq_name = ".buff.size"}
```
Comment (schweitzpgi) [inline flang/test/Lower/array-constructor-2.f90:108]: I'm confused why these compiler generated temps are now pinned. In fact, that seems incorrect. For example, if these were in a subprogram that was inlined into another subprogram, I'd expect these to be hoisted to the entry of the calling subprogram, etc.

**Review by schweitzpgi — COMMENTED** [review]
Looks like a number of spots with an undecorated `false` being added.

Is there some way to factor this a little finer such that a variable that requires pinning (such as for OMP privitization) gets it, but other compiler generated storage locations that have Fortran semantics do not?

File: flang/test/Lower/array-constructor-2.f90 (hunk 1 above)
Comment (kiranchandramohan) [reply flang/test/Lower/array-constructor-2.f90:108]: OK. I did not consider inlining. I will remove pinned for Fortran semantics.

**Comment by kiranchandramohan** [comment]
> Is there some way to factor this a little finer such that a variable that requires pinning (such as for OMP privitization) gets it, but other compiler generated storage locations that have Fortran semantics do not?

Restricted pinning to OpenMP privatisation.

**Review by schweitzpgi — APPROVED** [review]
Thanks, Kiran. +1

**Comment by kiranchandramohan** [comment]
Rebased. @schweitzpgi can this be merged? I will upstream the changes after it is merged.

**Pull request merged**